### PR TITLE
Feature/tamg

### PR DESCRIPTION
**Pull request by pohm01**
Add a matrix-free h-multigrid.

Currently only handles piecewise constant interpolation between grid levels.
Chebyshev as a smoother and coarse gird solver.

### CODE REVIEW

**Comment by timofeymukha** [comment]
Very cool stuff, Peter! 

Some quick comments

- Don't use `print`, we stick to `write`
- You may want to look at the suggestions of the linter under the develop/Flint/List chagend files check. At least sticking to col80 lines would be very nice.
- This is CPU-only, right? Do we have a check in the code?
- We should add the tamg to the documentation.
- Related to the above, when would one use this one as opposed to the hsmg. Is there an example where we could perhaps use it?

File: src/multigrid/tree_amg.f90 [hunk 1]
```diff
@@ -0,0 +1,330 @@
+! Copyright (c) 2020-2023, The Neko Authors
```
Comment (njansson) [inline src/multigrid/tree_amg.f90:1]:
```suggestion
! Copyright (c) 2024, The Neko Authors
```

File: src/multigrid/tree_amg_aggregate.f90 [hunk 2]
```diff
@@ -0,0 +1,672 @@
+! Copyright (c) 2020-2023, The Neko Authors
```
Comment (njansson) [inline src/multigrid/tree_amg_aggregate.f90:1]:
```suggestion
! Copyright (c) 2024, The Neko Authors
```

File: src/multigrid/tree_amg_multigrid.f90 [hunk 3]
```diff
@@ -0,0 +1,311 @@
+! Copyright (c) 2020-2023, The Neko Authors
```
Comment (njansson) [inline src/multigrid/tree_amg_multigrid.f90:1]:
```suggestion
! Copyright (c) 2024, The Neko Authors
```

File: src/multigrid/tree_amg_smoother.f90 [hunk 4]
```diff
@@ -0,0 +1,318 @@
+! Copyright (c) 2020-2023, The Neko Authors
```
Comment (njansson) [inline src/multigrid/tree_amg_smoother.f90:1]:
```suggestion
! Copyright (c) 2024, The Neko Authors
```

File: src/multigrid/tree_amg_utils.f90 [hunk 5]
```diff
@@ -0,0 +1,182 @@
+! Copyright (c) 2020-2023, The Neko Authors
```
Comment (njansson) [inline src/multigrid/tree_amg_utils.f90:1]:
```suggestion
! Copyright (c) 2024, The Neko Authors
```

**Review by njansson — COMMENTED** [review]
Some initial fixes for (C) year (the year should be when the file is added)

File: src/multigrid/phmg.f90 [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+                                     mg, intrp, msh, Ax, amg_solver)
+    type(ksp_monitor_t) :: ksp_results
+    integer :: lvl, clvl
+    type(phmg_lvl_t) :: mg(0:clvl)
+    type(interpolator_t) :: intrp(1:clvl)
+    type(tamg_solver_t), intent(inout) :: amg_solver
+    class(ax_t), intent(inout) :: Ax
+    type(mesh_t), intent(inout) :: msh
+    type(field_t) :: z, r, w
+    type(file_t) :: fname, cname
```
Comment (njansson) [inline src/multigrid/phmg.f90:225]:
```suggestion
```

**Review by njansson — APPROVED** [review]

**Pull request merged**